python/cuxfilter/charts/core/non_aggregate/core_graph.py

```diff
@@ -115,7 +115,12 @@ def __init__(
         self.edge_aggregate_col = edge_aggregate_col
         self.node_aggregate_fn = node_aggregate_fn
         self.edge_aggregate_fn = edge_aggregate_fn
-        self.node_color_palette = list(node_color_palette)
+        if node_color_palette is None:
+            self.no_colors_set = True
+            self.node_color_palette = CUXF_DEFAULT_COLOR_PALETTE
+        else:
+            self.node_color_palette = list(node_color_palette)
```

Comment:
Explicitly coordinating three values (parameter `node_color_palette`, the `no_colors_set` flag, and `self.node_color_palette`) in several places seems a bit fragile against future changes or additions. The amount of explicit state to manually set in the initializer could be reduced with some properties

```python
def __init__(self, ..., node_color_palette=None):
    self._node_color_palette_input = node_color_palette

@property
def colors_set(self):
    return self._node_color_palette_input is not None

@property 
def node_color_palette(self):
    if self.colors_set:
        return list(self._node_color_palette_input)
    return CUXF_DEFAULT_COLOR_PALETTE
```

Maybe not for now, but in the future it might be nice to partition the `node` and `edge` parts of this in to sub-objects. This palette handling code is almost-repeated below but can't be factored in to a mixin or base class easily because of the need to differentiate <pre>self.<b>node</b>_color_palette // self.<b>edge</b>_color_palette</pre> directly in the variable names. But things were split a little the palette naming could be generic via `self.node.color_palette` and `self.edge.color_palette`, and then all the `color_palette` handling could be factored out.

Reply:
fixed this